build.sbt

```diff
@@ -7,8 +7,9 @@ version := "0.0.1"
 scalaVersion := "2.12.11"
  
 libraryDependencies ++= Seq(
+  "com.tresys" % "dfdl-nitf-core" % "0.0.1",
```

Comment:
Thoughts on including both schemas in this same repo? That's likely what people that whant just plain nitf would expext. I'm thinking the directory structure might look something like this:

```
.
|-- README
|-- build.sbt
|-- nitf-core
|   |-- build.sbt
|    `-- src/main/resources/nitf.dfdl.xsd
`-- nitf
    |-- build.sbt
    `-- src/main/resources/com/tresys/nitf/xsd/nitf_extension_types.dfdl.xsd
```

So the nitf-core and nitf subdirectories can have a build.sbt in them that allow users to build or publish them individually if they run sbt from those directories. And the root can have a build.sbt that creates subprojects using those subdirectories, which allows easily running "sbt test" from the root to test everything. The subdirectory build.sbt files would look basically as they are now. And I think the root build.sbt would look something like this, so nothing too complicated:

```scala
lazy val root =     Project("root", file("."))
                      .aggregate(nitf, nitfCore)

lazy val nitf =     Project("nitf", file("nitf"))
                      .dependsOn(nitfCore)

lazy val nitfCore = Project("nitf-core", file("nitf-core"))
```

Reply:
The downside to that is if someone wants to parse plain nitf files, all the necessary files aren't in the jar. The nitf schema isn't complete without the `nitf_extension_types.dfdl.xsd` file on the classpath and providing the necessary types. So if it were moved to src/test/resoruces and someone wanted to parse plain nitf, they would need to pull out the file from src/test and put it on the classpath.

I was thinking of this sort of like `nitf-core` is an abstract class that requires some types/elements be defined. The `nitf` projects defines those types, but you could optionally defined those types differenting in the `raster` project.

And so nothing uses nitf-core alone, it's always a dependency to something that "extends" it by implementing those types.

Reply:
Is there such a thing as "plain NITF files" for real use, i.e., not just for test?

To my knowledge, there is always at least 1 image data type included. But I might be wrong. 

The "plain NITF file" without even JPEG in it, exists so we can do testing of just that aspect in isolation.

In my mind it's exactly analogous to NACT headers. They're never used alone. We can test them alone, but real usage always adds a payload. 

In the mil-std-2045 schema, we have a schema file, which must be found on the classpath, under src/test/resources, which defines the payload. NACT is similar. 

I guess this boils down to is NITF usable standalone, or are there always extensions?

Reply:
There's another technical subtlety here. If I include nitf_extension_types.dfdl.xsd in src/main/resources, then to create a schema for raster I have to control the order of Jars on the classpath so that the raster jar which redefines this file is earlier on the classpath. 

I'm not sure how one controls that classpath order in sbt, maven, etc.  I never thought of the library dependencies in sbt as being an ordered list, but perhaps it is?

Reply:
I think of "plain NITF" as [MIL-STD-2500C](https://gwg.nga.mil/ntb/baseline/docs/2500c/index.html), which does list data types that can be in the payload, e.g. JPEG and JPEG2000. That might not be correct though, and I'm not totally clear how it relates to Raster--I think Raster is just NITF with defined extensions?

Reply:
> There's another technical subtlety here. If I include nitf_extension_types.dfdl.xsd in src/main/resources, then to create a schema for raster I have to control the order of Jars on the classpath so that the raster jar which redefines this file is earlier on the classpath.

That's why I suggest splitting into nitf-core and nitf. nitf-core does not include nitf_extension_types.dfdl.xsd, but only imports it. Then we would have separate nitif and raster projects, each of which would depend on nitf-core and include a different nitf_extension_types.dfdl.xsd file. So there's only one extension types file on the classpath.

One issue with this approach is if you want to parse both NITF and Raster in the same process. Then you would have an issue where the extension types is on the classpath twice. But this seems like a limitation of XML Schema. This like overloading doesn't really seem possible.

Reply:
Then nitf-core does not support 'sbt test' because a required include/import file is missing?

The invariant that 'sbt test' always works and tests the module/assembly, that is super important.

Reply:
nitf-core could have its own implementation of the extension types in nitf-core/src/test that is only used for testing--the extension types could just be hexBinary blobs or something simple. Could also mean we could have separate projects like NITFwithJPEG, NITFwithJPEG2000, NITFwithEverything or something if we wanted. And they all could implement different extensions/payloads.

Or maybe another approach is needed. We really need someway to allow for raster to change parts of the nitf schema Splitting into separate nitf-core, nitf, and raster projects seemed like the only way to do it to me, but maybe there's an alternative.

Reply:
I found that the order of classpath is in fact the dependency order, that is if this stack-overflow response is believable:

https://stackoverflow.com/questions/33405025/sbt-modify-the-order-of-dependencies-in-the-classpath

So that would mean a schema can include things in separate files, and a jar earlier on the class path can override individual files to provide alternate definitions. 

I would suggest that in build.sbt when we rely on classpath order explicitly, we should say so in comments like
`// override nitf_extension_types.dfdl.xsd file`

Reply:
Ah, that's interesting. So if that's that case, we could keep nitf.dfdl.xsd exactly as it with only a single projet, and in order to do raster we just ensure the raster jar is on the classpath first. I'm a bit concerned that that's a bit fragile and could be easily messed up, since I don't think most people pay attention to classpath order, or even know about the sbt dependency order (I know I didn't).

Reply:
I agree that it's not like we have "protected" and "override" keywords to mark our intent, but I don't think this is conceptually too bad. Maybe this is what the XSD redefine element was supposed to be for. Alas that's not part of DFDL, only import and include.  

The other examples we have of this envelope-payload idiom aren't really overriding. E.g., the stanag-5516-with-nact schema depends on nact and stanag-5516 schemas, but the nact schema's own "payload" definition lives in src/test, so it's not really an override by the combined assembly schema, it's supplying a definition that would be missing. 

I suggest we try nitf as one nitf schema, which has the core and a nitf_extension_types.dfdl.xsd in src/master which  requires a dependency on JPEG, as we always have. Nitf's local tests can then still use jpeg images. 

Then the schema for raster (really nitf-with-raster) can depend on nitf and redefine/override the nitf_extension_types.dfdl.xsd file. 

Users of raster will have to insure they have the raster jar and the nitf jar, and the dependencies are in that order. I think that would happen naturally if someone expresses an application dependency on raster only, and the nitf dependency is found transitively.

Reply:
You guys may have mentioned this earlier, but the whole reason for the split was because I couldn't run daffodil from the command line to process Raster data.  Things worked just fine when testing through sbt, but I couldn't figure out a way to reliably process things on the command line.

Reply:
In theory, you should be able to do something like this:
```bash
export DAFFODIL_CLASSPATH="/path/to/raster.jar:/path/to/nitf.jar"
daffodil parse -s com/tresys/nitf/nitf.dfdl.xsd foo.nitf
```
So the raster jar must come first, followed by the nitf dependency. Then you use the classpath resource path found inside the nitf.jar as the root schema. When Daffodil compiles the schema and imports the extensions file, it should find the raster extension file first and use that.